server/superdesk/io/ingest_service.py

```diff
@@ -14,12 +14,62 @@
 from datetime import datetime
 from superdesk.utc import utc, utcnow
 from superdesk.errors import SuperdeskApiError, SuperdeskIngestError
+from superdesk.io import register_provider
 
 
 logger = logging.getLogger(__name__)
 
 
-class IngestService():
+class AutoRegisteredMeta(type):
+    """Metaclass for automatically registering the ingest service instances.
+
+    An instance of an ingest class is registered as a provider when the class
+    is instantiated - but only if the class defines a `PROVIDER` attribute
+    containing the name under which to register the class instances.
+
+    Every ingest provider class needs to define an `ERRORS` attribute
+    representing a list of <error_number, error_message> pairs that might be
+    raised by the class instances' methods.
+
+    If another provider class with the same `PROVIDER` name has already been
+    defined, a TypeError is raised.
+    """
+    _providers = {}  # keep track of all PROVIDER names defined in classes
+
+    def __new__(metacls, name, bases, attrs):
+        provider_name = attrs.get('PROVIDER')
+
+        if provider_name is not None:
+            if 'ERRORS' not in attrs:
+                raise AttributeError(
+                    "Provider class {} must define "
+                    "the ERRORS list attribute.".format(name))
+
+            if provider_name in metacls._providers:
+                raise TypeError(
+                    "PROVIDER {} already exists (class {}).".format(
+                        provider_name, metacls._providers[provider_name])
+                )
+
+            metacls._providers[provider_name] = name
+
+        return super().__new__(metacls, name, bases, attrs)
+
+    def __call__(cls, *args, **kwargs):
+        # NOTE: Registration cannot be done in __new__ or __init__, because the
+        # class cannot be instantiated there yet (a RuntimeError occurs), thus
+        # we need to perform the registration here in __call__ (when a new
+        # class instance is going to be created).
+
+        new_instance = super().__call__(*args, **kwargs)
+
+        if hasattr(cls, 'PROVIDER'):
+            register_provider(cls.PROVIDER, new_instance, cls.ERRORS)
```

Comment:
They should be instantiated at application startup?

The problem is that is not possible to create instances of the just-created class in `__new__()` or `__init__()` if that class invokes `super().__init__()` in its own `__init__()` method, a `RuntimeError` occurs. See [this discussion](http://stackoverflow.com/questions/13126727/how-is-super-in-python-3-implemented) for more details. This behavior is considered a bug in the Python interpreter by some people (I'm using CPython 3.3.6 on Ubuntu).

Before sending this PR, I circumvented the problem by registering the ingest classes on their instantiation, but if that's not OK, I can revert back to registering the classes in metaclass'es `__new__()` - if we can live with the limitation that `PROVIDER` ingest classes will not be able to call `super().__init__()` in their own `__init__()` method (FWIW, the former doesn't do anything really, since the base `IngestService` class doesn't have its own `__init__()` method).